File: benchmarks/lib/manifest.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { readFileSync } from "node:fs";
import {
  METRICS,
  loadManifest,
  casesForMetric,
  validateResultIds,
  resolveStreamingScript,
} from "./manifest.mjs";

const __dirname = dirname(fileURLToPath(import.meta.url));
const CORPUS = join(__dirname, "..", "corpus");

test("METRICS lists exactly the four metrics", () => {
  assert.deepEqual([...METRICS].sort(), ["cold-start", "first-run", "streaming", "warm"]);
});

test("loadManifest validates the committed corpus", () => {
  const m = loadManifest(CORPUS);
  assert.ok(m.cases.length >= 6, "expected at least 6 corpus cases");
  assert.ok(m.ids.has("trivial"));
});

test("casesForMetric filters by declared metric", () => {
  const m = loadManifest(CORPUS);
  const streaming = casesForMetric(m, "streaming");
  assert.ok(streaming.every((c) => c.metrics.includes("streaming")));
  assert.ok(streaming.length >= 1);
});

test("validateResultIds throws on an orphan id", () => {
  const m = loadManifest(CORPUS);
  assert.throws(() => validateResultIds(m, [{ id: "does-not-exist" }]), /orphan id/);
  assert.doesNotThrow(() => validateResultIds(m, [{ id: "trivial" }]));
});

test("resolveStreamingScript prefers streamingScript, falls back to script", () => {
  const m = loadManifest(CORPUS);
  const mapScale = m.cases.find((c) => c.id === "map-scale");
  const streamText = resolveStreamingScript(m, mapScale);
  assert.ok(streamText.includes("deferred=true"), "streaming variant declares deferred=true");

  const objTransform = m.cases.find((c) => c.id === "object-transform"); // no streamingScript
  const fallback = resolveStreamingScript(m, objTransform);
  const base = readFileSync(join(CORPUS, objTransform.script), "utf-8");
  assert.equal(fallback, base, "falls back to the base script when no streamingScript");
});

File: benchmarks/runners/engine/src/test/scala/org/mule/weave/benchmark/engine/ManifestTest.scala
package org.mule.weave.benchmark.engine

import org.scalatest.freespec.AnyFreeSpec
import org.scalatest.matchers.should.Matchers
import java.io.File

class ManifestTest extends AnyFreeSpec with Matchers {
  private val corpus = new File("../../corpus").getCanonicalFile

  "loads all corpus cases with stable ids" in {
    val m = Manifest.load(corpus)
    m.ids should contain allOf ("trivial", "object-transform", "map-scale", "xml-to-csv", "json-stream", "compile-heavy")
  }

  "casesForMetric filters by declared metrics" in {
    val m = Manifest.load(corpus)
    Manifest.casesForMetric(m, "streaming").map(_.id) should contain allOf ("map-scale", "json-stream")
    Manifest.casesForMetric(m, "cold-start").map(_.id) should contain ("trivial")
  }

  "resolveScript reads the .dwl source" in {
    val m = Manifest.load(corpus)
    val trivial = m.cases.find(_.id == "trivial").get
    Manifest.resolveScript(m, trivial).trim should include ("2 + 2")
  }

  "resolveInputs reads committed input bytes with mime + charset" in {
    val m = Manifest.load(corpus)
    val xml = m.cases.find(_.id == "xml-to-csv").get
    val inputs = Manifest.resolveInputs(m, xml)
    inputs should have size 1
    inputs.head.name shouldBe "payload"
    inputs.head.mimeType shouldBe "application/xml"
    inputs.head.charset shouldBe Some("UTF-16")
    inputs.head.bytes.length should be > 0
  }

  "validateResultIds throws on an orphan id" in {
    val m = Manifest.load(corpus)
    an [RuntimeException] should be thrownBy Manifest.validateResultIds(m, Seq("trivial", "not-a-case"))
    noException should be thrownBy Manifest.validateResultIds(m, Seq("trivial"))
  }

  "resolveStreamingScript prefers the streamingScript variant" in {
    val corpus = new File("../../corpus").getCanonicalFile
    val m = Manifest.load(corpus)
    val mapScale = m.cases.find(_.id == "map-scale").get
    Manifest.resolveStreamingScript(m, mapScale) should include ("deferred=true")
  }

  "resolveStreamingScript falls back to the base script" in {
    val corpus = new File("../../corpus").getCanonicalFile
    val m = Manifest.load(corpus)
    val obj = m.cases.find(_.id == "object-transform").get // no streamingScript
    Manifest.resolveStreamingScript(m, obj) shouldBe Manifest.resolveScript(m, obj)
  }
}
